app/services/lms_sync_service.py

```diff
@@ -0,0 +1,241 @@
+        })
+        db_students = await self.student_service.list_students()
+        
+        if(db_students is not None):
+            # Delete students that are in the database but not in Canvas
+            for student in db_students:
+                student_pid = await self.canvas_service.get_pid_from_onyen(student.onyen)
+                if student_pid not in [s['sis_user_id'] for s in canvas_students]:
+                    await self.canvas_service.unassociate_pid_from_user(student.onyen)
+                    await self.student_service.delete_user(student.onyen)
+       
+        for student in canvas_students:
+            pid = student['sis_user_id']
+            user_info = self.ldap_service.get_user_info(pid)
+
+            try:
+                await self.student_service.get_user_by_onyen(user_info.onyen)
+
+            except UserNotFoundException:
+                #create a new student
+                await self.student_service.create_student(
+                    onyen=user_info.onyen,
+                    name=student['name'],
+                    email=student['email']
+                )
+                await self.canvas_service.associate_pid_to_user(user_info.onyen, pid)
+
+        return canvas_students
+    
+    async def sync_instructors(self):
+        canvas_instructors = await self.canvas_service.get_users({
+            "enrollment_type": "teacher"
+        })
+        db_instructors = await self.instructor_service.list_instructors()
+       
+        if(db_instructors is not None):       
+            # Delete instructors that are in the database but not in Canvas
+            for instructor in db_instructors:
+                instructor_pid = await self.canvas_service.get_pid_from_onyen(instructor.onyen)
+                if instructor_pid not in [i['sis_user_id'] for i in canvas_instructors]:
+                    await self.canvas_service.unassociate_pid_from_user(instructor.onyen)
+                    await self.instructor_service.delete_user(instructor.onyen)
+        
+        for instructor in canvas_instructors:
+            pid = instructor['sis_user_id']
+            user_info = self.ldap_service.get_user_info(pid)
+
+            try:
+                await self.instructor_service.get_user_by_onyen(user_info.onyen)
+
+            except UserNotFoundException:
+                #create a new instructor
+                await self.instructor_service.create_instructor(
+                    onyen=user_info.onyen,
+                    name=instructor['name'],
+                    email=instructor['email']
+                )
+                await self.canvas_service.associate_pid_to_user(user_info.onyen, pid)
+
+        return canvas_instructors
+
+    async def upload_grades_from_csv(self, grade_csv):
+        df = pd.read_csv(io.StringIO(grade_csv.decode("utf-8")))
+        
+        for index, row in df.iterrows():
+            try:
+                assignment_id = self._extract_id(row['file'])
```

Comment:
This is bad. `assignment_id` should 100% be an argument of this method.

Reply:
done